eth/api.go

```diff
@@ -1427,6 +1427,143 @@ func (api *PrivateDebugAPI) SetHead(number uint64) {
 	api.eth.BlockChain().SetHead(number)
 }
 
+// StructLogRes stores a structured log emitted by the evm while replaying a
+// transaction in debug mode
+type structLogRes struct {
+	Pc      uint64            `json:"pc"`
+	Op      string            `json:"op"`
+	Gas     *big.Int          `json:"gas"`
+	GasCost *big.Int          `json:"gasCost"`
+	Error   error             `json:"error"`
+	Stack   []string          `json:"stack"`
+	Memory  map[string]string `json:"memory"`
+	Storage map[string]string `json:"storage"`
+}
+
+// TransactionExecutionRes groups all structured logs emitted by the evm
+// while replaying a transaction in debug mode as well as the amount of
+// gas used and the return value
+type TransactionExecutionRes struct {
+	Gas         *big.Int       `json:"gas"`
+	ReturnValue string         `json:"returnValue"`
+	StructLogs  []structLogRes `json:"structLogs"`
+}
+
+func (s *PrivateDebugAPI) doReplayTransaction(txHash common.Hash) ([]vm.StructLog, []byte, *big.Int, error) {
+	// Retrieve the tx from the chain
+	tx, _, blockIndex, _ := core.GetTransaction(s.eth.ChainDb(), txHash)
+
+	if tx == nil {
+		return nil, nil, nil, fmt.Errorf("Transaction not found")
+	}
+
+	block := s.eth.BlockChain().GetBlockByNumber(blockIndex - 1)
+	if block == nil {
+		return nil, nil, nil, fmt.Errorf("Unable to retrieve prior block")
+	}
+
+	// Create the state database
+	stateDb, err := state.New(block.Root(), s.eth.ChainDb())
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	txFrom, err := tx.From()
+
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("Unable to create transaction sender")
+	}
+	from := stateDb.GetOrNewStateObject(txFrom)
+	msg := callmsg{
+		from:     from,
+		to:       tx.To(),
+		gas:      tx.Gas(),
+		gasPrice: tx.GasPrice(),
+		value:    tx.Value(),
+		data:     tx.Data(),
+	}
+
+	vmenv := core.NewEnv(stateDb, s.eth.BlockChain(), msg, block.Header())
+	gp := new(core.GasPool).AddGas(block.GasLimit())
+	vm.GenerateStructLogs = true
+	defer func() { vm.GenerateStructLogs = false }()
+
+	ret, gas, err := core.ApplyMessage(vmenv, msg, gp)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("Error executing transaction %v", err)
+	}
+
+	return vmenv.StructLogs(), ret, gas, nil
+}
+
+func (s *PrivateDebugAPI) ReplayTransaction(txHash common.Hash, stackDepth int, memorySize int, storageSize int) (TransactionExecutionRes, error) {
```

Comment:
Minor issues:

 * Please add a method doc as it's a public method
 * Returning a pointer to a `TransactionExecutionRes` is better, as you can return nil in case of errors instead f having to construct empty objects and copy them over the place.

Also please name it `TransactionExecutionResult` (i.e. it's s large name already, that 3 characters from the end won't save you any significant typing :P )